### PR TITLE
[FEATURE] clip gradient norm by parameter

### PR DESCRIPTION
**Pull request by szha**
## Description ##
clip gradient norm by parameter

### Essentials ###
- [x] PR's title starts with a category (e.g. [BUGFIX], [MODEL], [TUTORIAL], [FEATURE], [DOC], etc)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [x] Code is well-documented

### Changes ###
- [x] function for clipping gradient norm given parameters

## Comments ##
- clipping gradient norms for multi-device training is hard to get right with the existing clip_global_norm, because gradients should be added together first.

### CODE REVIEW

**Comment by mli** [comment]
Job PR-470/3 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-470/3/index.html

File: src/gluonnlp/utils/parameter.py [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+    def _reduce_grad(parameter):
+        """Sum gradients of all devices onto the first context and return gradient sum.
+        This function has side effect and does not preserve the original per-device gradients."""
+        grads = parameter.list_grad()
+        if len(grads) > 1:
+            for s in reversed(range(len(grads))):
+                t = int(s/2)
+                if s != t:
+                    target = grads[t]
+                    target[:] = target + grads[s].as_in_context(target.context)
```
Comment (eric-haibin-lin) [inline src/gluonnlp/utils/parameter.py]: use `target += grads[s].as_in_context(target.context)` to reduce a temp copy

File: src/gluonnlp/utils/parameter.py [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+                    continue
+        return grads[0]
+
+    def _broadcast_grad(parameter):
+        """Broadcast the gradient on the first context to the copies on all other contexts."""
+        grads = parameter.list_grad()
+        if len(grads) > 1:
+            for t in range(1, len(grads)):
+                target = grads[t]
+                target[:] = grads[0].as_in_context(target.context)
```
Comment (eric-haibin-lin) [inline src/gluonnlp/utils/parameter.py]: use `grads[0].copyto(target)` to reduce a temp copy

File: src/gluonnlp/utils/parameter.py [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+    ctx = arrays[0].context
+    total_norm = nd.add_n(*[_norm(arr).as_in_context(ctx) for arr in arrays])
+    total_norm = nd.sqrt(total_norm)
+    if check_isfinite:
+        if not np.isfinite(total_norm.asscalar()):
+            warnings.warn(
+                UserWarning('nan or inf is detected. '
+                            'Clipping results will be undefined.'), stacklevel=2)
+    scale = max_norm / (total_norm + 1e-8)
+    scale = nd.min(nd.concat(scale, nd.ones(1, ctx=ctx), dim=0))
```
Comment (eric-haibin-lin) [inline src/gluonnlp/utils/parameter.py]: pass scale.dtype to nd.ones to preserve the dtype in case for fp16

File: src/gluonnlp/utils/parameter.py [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+    for arr in arrays:
+        arr[:] = arr * scale.as_in_context(arr.context)
```
Comment (eric-haibin-lin) [inline src/gluonnlp/utils/parameter.py]: use `arr*= scale.as_in_ctx()` to reduce a temp copy

File: src/gluonnlp/utils/parameter.py [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+            return nd.dot(x, x)
+        return array.norm().square()
+
+    arrays = [_reduce_grad(p) for p in parameters if p.grad_req != 'null']
+    assert len(arrays) > 0
+    ctx = arrays[0].context
+    total_norm = nd.add_n(*[_norm(arr).as_in_context(ctx) for arr in arrays])
+    total_norm = nd.sqrt(total_norm)
+    if check_isfinite:
+        if not np.isfinite(total_norm.asscalar()):
```
Comment (eric-haibin-lin) [inline src/gluonnlp/utils/parameter.py]: We can assign `total_norm = total_norm.asscalar()` if check_isfinite so that it is only done once, then you don't need the if stmt at the end

**Comment by szha** [comment]
I had an offline discussion with @eric-haibin-lin and it seems suboptimal to include the allreduce logic in this function since it's already part of the trainer's allreduce_grads function. I will update the new function to exclude the reduce logic, and also reduce communication.

**Comment by mli** [comment]
Job PR-470/5 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-470/5/index.html

File: src/gluonnlp/utils/parameter.py [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+        total_norm = total_norm.asscalar()
+        if not np.isfinite(total_norm):
+            warnings.warn(
+                UserWarning('nan or inf is detected. '
+                            'Clipping results will be undefined.'), stacklevel=2)
+    scale = max_norm / (total_norm + 1e-8)
+    if check_isfinite:
+        scale = nd.array([scale], dtype=dtype, ctx=ctx)
+    scale = nd.min(nd.concat(scale, nd.ones((1,), dtype=dtype, ctx=ctx), dim=0))
+    for p in parameters:
```
Comment (eric-haibin-lin) [inline src/gluonnlp/utils/parameter.py]: Why not use `arrays` directly for the for loop?

Comment (szha) [reply src/gluonnlp/utils/parameter.py]: `arrays` only has the gradient array of the first context for each parameter.
